### PR TITLE
Bugfix of model.export() to work correct with bs>1

### PR DESCRIPTION
**Pull request by BloodAxe**
1) Picking wrong indexes with exporting models with bs>1 in decoding modules (YoloNAS, YoloNAS-Pose, YoloX, PPYolo-E)

 2) When using ONNX NonMaxSupression a max_predictions_per_image parameter was used incorrectly (Flat/Batch format predictions could contain more than a requested number of detections)

### CODE REVIEW

**Review by shaydeci — CHANGES_REQUESTED** [review]

File: src/super_gradients/conversion/onnx/nms.py [hunk 1]
```diff
@@ -32,49 +44,78 @@ def __init__(self, batch_size: int, num_pre_nms_predictions: int, max_prediction
...
 
-        for i in range(self.batch_size):
-            selected_predictions = predictions[masks[i]]
-            selected_predictions = selected_predictions[:, 1:]
-            batch_predictions[i] = torch.nn.functional.pad(
-                selected_predictions, (0, 0, 0, self.max_predictions_per_image - selected_predictions.size(0)), value=0, mode="constant"
-            )
+            # Add dummy row to mask and predictions to ensure that we always have at least one prediction per image
+            # ONNX/TRT deals poorly with tensors that has zero dims, and we need to ensure that we always have at least one prediction per image
+            masks = torch.cat([masks, torch.zeros((self.batch_size, 1), dtype=masks.dtype, device=predictions.device)], dim=1)  # [B, L+1]
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/nms.py:87]: The problem we are facing with TRT is some sort of edge case when you have tensor of shape `[Batch, Num Detections]`. when `Num Detections = 0`.
Empty predictions may appear when there is no detections in batch. 

And for some reason even basic constructions (`F.pad` or `view(-1)` and `squeeze(1)` do not work properly. Even `torch.sum` fails when input has `[B, 0]` shape). The only workaround I was able to find to combat this scenario is `torch.cat`. 

So the intention is indeed correct - adding a False/Zero items to ensure that shape of masks/preds has at least one element in Num Detections. Since mask padded with zeros - it does not affect existing predictions. 

In the nutshell - this is a dirty trick to fool TRT and make it do what we want him to do 🙌

File: src/super_gradients/conversion/onnx/nms.py [hunk 2]
```diff
@@ -244,7 +343,7 @@ def attach_onnx_nms(
     graph.layer(op="Transpose", name="permute_scores", inputs=[pred_scores], outputs=[permute_scores], attrs={"perm": [0, 2, 1]})
 
     op_inputs = [pred_boxes, permute_scores] + [
-        gs.Constant(name="max_output_boxes_per_class", values=np.array([max_predictions_per_image], dtype=np.int64)),
+        gs.Constant(name="max_output_boxes_per_class", values=np.array([num_pre_nms_predictions], dtype=np.int64)),
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/nms.py:346]: ONNX NMS does not output boxes / labels / scores directly. Instead it outputs indexes of **input** predictions that we want to keep. And the length of those input predictions is `max_predictions_per_image`.

File: src/super_gradients/conversion/onnx/nms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/conversion/onnx/nms.py:87]: I understand.
But in the case where num_detections=0, we add dummy row and another dummy row to the masks = False.
Then we won't select the dummy row, and still end up with a size=0 in the last dimension.
Or is it ok from that point on?

Maybe I missed it - but I didn't see we test export works in the case of no predictions.
Can we add such?

[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/nms.py:87]: Gather by mask when mask has shape [B,0] is somehow works fine. It is mainly for num_predictions = torch.sum(masks, dim=1, keepdim=True).long()

**Review by Louis-Dupont — COMMENTED** [review]
LGTM overall, I just didn't dig super deep into each line of code, I'll trust your tests and @shaydeci this time 😇 
Let me know if there is something specific you want me to review in detail
